File: tauri/src/lib/services/systemApi.ts
import { backend, type Profile } from "./backend.svelte";
import { http, type HttpClient } from "./http";
import { ReconnectingSocket } from "./socket";

export interface DirEntry {
  name: string;
  path: string;
  isDir: boolean;
}

export interface DirListing {
  path: string;
  parent: string | null;
  roots: string[];
  entries: DirEntry[];
}

export interface DiskInfo {
  mount: string;
  used: number;
  total: number;
  percent: number;
}

export interface GpuInfo {
  name: string;
  percent: number;
  memUsed: number;
  memTotal: number;
  memPercent: number;
  temp: number | null;
}

export interface BatteryInfo {
  percent: number;
  plugged: boolean;
  secsLeft: number | null;
}

export interface SystemInfo {
  hostname: string;
  os: string;
  osId: string;
  arch: string;
  cpuName: string | null;
  uptime: number;
  cpuPercent: number;
  cpuCores: number;
  memoryUsed: number;
  memoryTotal: number;
  memoryPercent: number;
  gpu: GpuInfo | null;
  battery: BatteryInfo | null;
  disks: DiskInfo[];
  netRx: number;
  netTx: number;
}

export interface LogEntry {
  ts: number;
  level: string;
  message: string;
}

export interface SystemHandlers {
  onInfo: (info: SystemInfo) => void;
  onLogs: (items: LogEntry[]) => void;
  onDrop: () => void;
}

type Wire = Record<string, any>;

const PING_SECONDS = 15;

const parseInfo = (raw: Wire): SystemInfo => ({
  hostname: raw.hostname ?? "",
  os: raw.os ?? "",
  osId: raw.os_id ?? "",
  arch: raw.arch ?? "",
  cpuName: raw.cpu_name ?? null,
  uptime: raw.uptime ?? 0,
  cpuPercent: raw.cpu?.percent ?? 0,
  cpuCores: raw.cpu?.cores ?? 0,
  memoryUsed: raw.memory?.used ?? 0,
  memoryTotal: raw.memory?.total ?? 0,
  memoryPercent: raw.memory?.percent ?? 0,
  gpu: raw.gpu
    ? {
        name: raw.gpu.name ?? "",
        percent: raw.gpu.percent ?? 0,
        memUsed: raw.gpu.mem_used ?? 0,
        memTotal: raw.gpu.mem_total ?? 0,
        memPercent: raw.gpu.mem_percent ?? 0,
        temp: raw.gpu.temp ?? null,
      }
    : null,
  battery: raw.battery
    ? {
        percent: raw.battery.percent ?? 0,
        plugged: raw.battery.plugged === true,
        secsLeft: raw.battery.secsleft ?? null,
      }
    : null,
  disks: (raw.disks ?? []).map((disk: Wire) => ({
    mount: disk.mount ?? "",
    used: disk.used ?? 0,
    total: disk.total ?? 0,
    percent: disk.percent ?? 0,
  })),
  netRx: raw.network?.rx ?? 0,
  netTx: raw.network?.tx ?? 0,
});

const parseLog = (raw: Wire): LogEntry => ({
  ts: raw.ts ?? 0,
  level: raw.level ?? "",
  message: raw.message ?? "",
});

export const createSystemApi = (client: HttpClient, profile: () => Profile) => ({
  async restart(): Promise<boolean> {
    return (await client.post("/system/restart")) !== null;
  },

  async dirs(path: string, files = false): Promise<DirListing | null> {
    const data = await client.get<Wire>(`/system/dirs?path=${encodeURIComponent(path)}&files=${files}`);
    if (!data) return null;
    return {
      path: typeof data.path === "string" ? data.path : "",
      parent: typeof data.parent === "string" ? data.parent : null,
      roots: Array.isArray(data.roots) ? (data.roots as string[]) : [],
      entries: (Array.isArray(data.entries) ? (data.entries as Wire[]) : []).map((item) => ({
        name: typeof item.name === "string" ? item.name : "",
        path: typeof item.path === "string" ? item.path : "",
        isDir: item.is_dir === true,
      })),
    };
  },

  stream(handlers: SystemHandlers): ReconnectingSocket {
    const socket = new ReconnectingSocket(
      "/system/ws",
      {
        onMessage: (message) => {
          if (message.type === "system") handlers.onInfo(parseInfo(message));
          else if (message.type === "logs") handlers.onLogs(((message.items as Wire[]) ?? []).map(parseLog));
        },
        onDrop: handlers.onDrop,
      },
      profile,
      PING_SECONDS,
    );
    socket.connect();
    return socket;
  },
});

export const systemApi = createSystemApi(http, () => backend.active);

File: client/app/src/commonMain/kotlin/com/jahirtrap/cconnect/data/remote/SystemApi.kt
package com.jahirtrap.cconnect.data.remote

import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.floatOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.longOrNull

object SystemApi {

    data class DirEntry(val name: String, val path: String, val isDir: Boolean)

    data class DirListing(
        val path: String,
        val parent: String?,
        val roots: List<String>,
        val entries: List<DirEntry>,
    )

    data class DiskInfo(
        val mount: String,
        val used: Long,
        val total: Long,
        val percent: Float,
    )

    data class GpuInfo(
        val name: String,
        val percent: Float,
        val memUsed: Long,
        val memTotal: Long,
        val memPercent: Float,
        val temp: Int?,
    )

    data class BatteryInfo(
        val percent: Float,
        val plugged: Boolean,
        val secsLeft: Long?,
    )

    data class SystemInfo(
        val hostname: String,
        val os: String,
        val osId: String,
        val arch: String,
        val cpuName: String?,
        val uptime: Double,
        val cpuPercent: Float,
        val cpuCores: Int,
        val memoryUsed: Long,
        val memoryTotal: Long,
        val memoryPercent: Float,
        val gpu: GpuInfo?,
        val battery: BatteryInfo?,
        val disks: List<DiskInfo>,
        val netRx: Double = 0.0,
        val netTx: Double = 0.0,
    )

    data class LogEntry(
        val ts: Double,
        val level: String,
        val message: String,
    )

    sealed interface Event {
        data class Info(val info: SystemInfo) : Event
        data class Logs(val items: List<LogEntry>) : Event
    }

    suspend fun restart(): Boolean = Http.post("/system/restart", buildJsonObject {}) != null

    /** Folder tree for the path picker, for the platforms without a native file dialog. */
    suspend fun dirs(path: String, files: Boolean = false): DirListing? {
        val data = Http.get("/system/dirs", mapOf("path" to path, "files" to files.toString()))?.jsonObject
            ?: return null
        return DirListing(
            path = data["path"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            parent = data["parent"]?.jsonPrimitive?.contentOrNull,
            roots = data["roots"]?.jsonArray?.mapNotNull { it.jsonPrimitive.contentOrNull }.orEmpty(),
            entries = data["entries"]?.jsonArray?.map { element ->
                val entry = element.jsonObject
                DirEntry(
                    name = entry["name"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    path = entry["path"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    isDir = entry["is_dir"]?.jsonPrimitive?.contentOrNull == "true",
                )
            }.orEmpty(),
        )
    }

    fun stream(): Flow<Event> = callbackFlow {
        val socket = openWebSocket(Backend.systemWsUrl, Backend.authHeaders, object : WsListener {
            override fun onOpen() {}

            override fun onMessage(text: String) {
                val o = runCatching { Json.parseToJsonElement(text).jsonObject }.getOrNull() ?: return
                when (o["type"]?.jsonPrimitive?.contentOrNull) {
                    "system" -> trySend(Event.Info(parseInfo(o)))
                    "logs" -> trySend(Event.Logs(o["items"]?.jsonArray?.map { parseLog(it.jsonObject) }.orEmpty()))
                }
            }

            override fun onFailure(reason: String) {
                close(RuntimeException(reason))
            }

            override fun onClosed(reason: String) {
                close()
            }
        }, pingSeconds = 15)
        awaitClose { socket.cancel() }
    }

    private fun parseInfo(o: JsonObject): SystemInfo {
        val cpu = o["cpu"]?.jsonObject
        val memory = o["memory"]?.jsonObject
        return SystemInfo(
            hostname = o["hostname"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            os = o["os"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            osId = o["os_id"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            arch = o["arch"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            cpuName = o["cpu_name"]?.jsonPrimitive?.contentOrNull,
            uptime = o["uptime"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
            cpuPercent = cpu?.get("percent")?.jsonPrimitive?.floatOrNull ?: 0f,
            cpuCores = cpu?.get("cores")?.jsonPrimitive?.intOrNull ?: 0,
            memoryUsed = memory?.get("used")?.jsonPrimitive?.longOrNull ?: 0L,
            memoryTotal = memory?.get("total")?.jsonPrimitive?.longOrNull ?: 0L,
            memoryPercent = memory?.get("percent")?.jsonPrimitive?.floatOrNull ?: 0f,
            gpu = (o["gpu"] as? JsonObject)?.let { gpu ->
                GpuInfo(
                    name = gpu["name"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    percent = gpu["percent"]?.jsonPrimitive?.floatOrNull ?: 0f,
                    memUsed = gpu["mem_used"]?.jsonPrimitive?.longOrNull ?: 0L,
                    memTotal = gpu["mem_total"]?.jsonPrimitive?.longOrNull ?: 0L,
                    memPercent = gpu["mem_percent"]?.jsonPrimitive?.floatOrNull ?: 0f,
                    temp = gpu["temp"]?.jsonPrimitive?.intOrNull,
                )
            },
            battery = (o["battery"] as? JsonObject)?.let { battery ->
                BatteryInfo(
                    percent = battery["percent"]?.jsonPrimitive?.floatOrNull ?: 0f,
                    plugged = battery["plugged"]?.jsonPrimitive?.contentOrNull == "true",
                    secsLeft = battery["secsleft"]?.jsonPrimitive?.longOrNull,
                )
            },
            disks = o["disks"]?.jsonArray?.map { el ->
                val d = el.jsonObject
                DiskInfo(
                    mount = d["mount"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    used = d["used"]?.jsonPrimitive?.longOrNull ?: 0L,
                    total = d["total"]?.jsonPrimitive?.longOrNull ?: 0L,
                    percent = d["percent"]?.jsonPrimitive?.floatOrNull ?: 0f,
                )
            }.orEmpty(),
            netRx = o["network"]?.jsonObject?.get("rx")?.jsonPrimitive?.doubleOrNull ?: 0.0,
            netTx = o["network"]?.jsonObject?.get("tx")?.jsonPrimitive?.doubleOrNull ?: 0.0,
        )
    }

    private fun parseLog(o: JsonObject): LogEntry = LogEntry(
        ts = o["ts"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
        level = o["level"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        message = o["message"]?.jsonPrimitive?.contentOrNull.orEmpty(),
    )
}
